### ISSUE TITLE
Binding failures with @ConfigurationProperties on Weblogic 

### ISSUE THREAD

**Issue by aritzbastida** [spring-projects/spring-boot#12408]
I have a Spring Boot (2.0.0.RELEASE) app which must be portable: it must work either standalone (as an executable archive) or deployed on Weblogic. As an executable archive, it starts OK, but on Weblogic I have ClassNotFoundException in the binding of @ConfigurationProperties, namely:
- **JpaProperties** (spring-boot-autoconfigure):  It tries to load org.hibernate.boot.model.naming.ImplicitNamingStrategy class (which comes from hibernate-core, but we are excluding this module, as we are using Eclipselink instead). 
- **KeyProperties** (spring-cloud-context / EncryptionBootstrapConfiguration): It tries to load org.springframework.security.rsa.crypto.RsaAlgorithm (which exists on the optional spring-security-rsa module; but we are not including it as dependency).

The exception raises on JavaBeanBinder.putProperties(JavaBeanBinder.java:108), which tries to call getDeclaredMethods() on these two classes... and fails with CNFE. That seems even logical, but then... why is it working OK in standalone mode? I have no clue. 

All I got to see is that, in standalone, it doesn't call JavaBeanBinder.putProperties() for KeyProperties; but in Weblogic, it does... Any help is appreciated. I attach a stacktrace:

```
org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'encryptionBootstrapConfiguration': Unsatisfied dependency expressed through field 'key'; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'encrypt-org.springframework.cloud.bootstrap.encrypt.KeyProperties': Initialization of bean failed; nested exception is java.lang.NoClassDefFoundError: org/springframework/security/rsa/crypto/RsaAlgorithm
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredFieldElement.inject(AutowiredAnnotationBeanPostProcessor.java:587)
	at org.springframework.beans.factory.annotation.InjectionMetadata.inject(InjectionMetadata.java:91)
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor.postProcessPropertyValues(AutowiredAnnotationBeanPostProcessor.java:373)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.populateBean(AbstractAutowireCapableBeanFactory.java:1344)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:582)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:502)
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:312)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:228)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:310)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:200)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:760)
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:868)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:549)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:752)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:388)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:327)
	at org.springframework.boot.builder.SpringApplicationBuilder.run(SpringApplicationBuilder.java:137)
	at org.springframework.cloud.bootstrap.BootstrapApplicationListener.bootstrapServiceContext(BootstrapApplicationListener.java:197)
	at org.springframework.cloud.bootstrap.BootstrapApplicationListener.onApplicationEvent(BootstrapApplicationListener.java:104)
	at org.springframework.cloud.bootstrap.BootstrapApplicationListener.onApplicationEvent(BootstrapApplicationListener.java:70)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.doInvokeListener(SimpleApplicationEventMulticaster.java:172)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.invokeListener(SimpleApplicationEventMulticaster.java:165)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:139)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:127)
	at org.springframework.boot.context.event.EventPublishingRunListener.environmentPrepared(EventPublishingRunListener.java:74)
	at org.springframework.boot.SpringApplicationRunListeners.environmentPrepared(SpringApplicationRunListeners.java:54)
	at org.springframework.boot.SpringApplication.prepareEnvironment(SpringApplication.java:351)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:317)
	at org.springframework.boot.web.servlet.support.SpringBootServletInitializer.run(SpringBootServletInitializer.java:155)
	at org.springframework.boot.web.servlet.support.SpringBootServletInitializer.createRootApplicationContext(SpringBootServletInitializer.java:135)
	at org.springframework.boot.web.servlet.support.SpringBootServletInitializer.onStartup(SpringBootServletInitializer.java:87)
	at org.springframework.web.SpringServletContainerInitializer.onStartup(SpringServletContainerInitializer.java:172)
	at weblogic.servlet.internal.WebAppServletContext.initContainerInitializer(WebAppServletContext.java:1420)
	at weblogic.servlet.internal.WebAppServletContext.initContainerInitializers(WebAppServletContext.java:1359)
	at weblogic.servlet.internal.WebAppServletContext.initContainerInitializers(WebAppServletContext.java:1340)
	at weblogic.servlet.internal.WebAppServletContext.preloadResources(WebAppServletContext.java:1906)
	at weblogic.servlet.internal.WebAppServletContext.start(WebAppServletContext.java:3063)
	at weblogic.servlet.internal.WebAppModule.startContexts(WebAppModule.java:1830)
	at weblogic.servlet.internal.WebAppModule.start(WebAppModule.java:875)
	at weblogic.application.internal.ExtensibleModuleWrapper$StartStateChange.next(ExtensibleModuleWrapper.java:360)
	at weblogic.application.internal.ExtensibleModuleWrapper$StartStateChange.next(ExtensibleModuleWrapper.java:356)
	at weblogic.application.utils.StateMachineDriver.nextState(StateMachineDriver.java:45)
	at weblogic.application.internal.ExtensibleModuleWrapper.start(ExtensibleModuleWrapper.java:138)
	at weblogic.application.internal.flow.ModuleListenerInvoker.start(ModuleListenerInvoker.java:124)
	at weblogic.application.internal.flow.ModuleStateDriver$3.next(ModuleStateDriver.java:237)
	at weblogic.application.internal.flow.ModuleStateDriver$3.next(ModuleStateDriver.java:232)
	at weblogic.application.utils.StateMachineDriver.nextState(StateMachineDriver.java:45)
	at weblogic.application.internal.flow.ModuleStateDriver.start(ModuleStateDriver.java:82)
	at weblogic.application.internal.flow.StartModulesFlow.activate(StartModulesFlow.java:52)
	at weblogic.application.internal.BaseDeployment$2.next(BaseDeployment.java:753)
	at weblogic.application.utils.StateMachineDriver.nextState(StateMachineDriver.java:45)
	at weblogic.application.internal.BaseDeployment.activate(BaseDeployment.java:263)
	at weblogic.application.internal.SingleModuleDeployment.activate(SingleModuleDeployment.java:53)
	at weblogic.application.internal.DeploymentStateChecker.activate(DeploymentStateChecker.java:165)
	at weblogic.deploy.internal.targetserver.AppContainerInvoker.activate(AppContainerInvoker.java:80)
	at weblogic.deploy.internal.targetserver.operations.AbstractOperation.activate(AbstractOperation.java:601)
	at weblogic.deploy.internal.targetserver.operations.ActivateOperation.activateDeployment(ActivateOperation.java:171)
	at weblogic.deploy.internal.targetserver.operations.ActivateOperation.doCommit(ActivateOperation.java:121)
	at weblogic.deploy.internal.targetserver.operations.AbstractOperation.commit(AbstractOperation.java:343)
	at weblogic.deploy.internal.targetserver.DeploymentManager.handleDeploymentCommit(DeploymentManager.java:895)
	at weblogic.deploy.internal.targetserver.DeploymentManager.activateDeploymentList(DeploymentManager.java:1422)
	at weblogic.deploy.internal.targetserver.DeploymentManager.handleCommit(DeploymentManager.java:454)
	at weblogic.deploy.internal.targetserver.DeploymentServiceDispatcher.commit(DeploymentServiceDispatcher.java:181)
	at weblogic.deploy.service.internal.targetserver.DeploymentReceiverCallbackDeliverer.doCommitCallback(DeploymentReceiverCallbackDeliverer.java:195)
	at weblogic.deploy.service.internal.targetserver.DeploymentReceiverCallbackDeliverer.access$100(DeploymentReceiverCallbackDeliverer.java:13)
	at weblogic.deploy.service.internal.targetserver.DeploymentReceiverCallbackDeliverer$2.run(DeploymentReceiverCallbackDeliverer.java:68)
	at weblogic.work.SelfTuningWorkManagerImpl$WorkAdapterImpl.run(SelfTuningWorkManagerImpl.java:643)
	at weblogic.invocation.ComponentInvocationContextManager._runAs(ComponentInvocationContextManager.java:348)
	at weblogic.invocation.ComponentInvocationContextManager.runAs(ComponentInvocationContextManager.java:333)
	at weblogic.work.LivePartitionUtility.doRunWorkUnderContext(LivePartitionUtility.java:54)
	at weblogic.work.PartitionUtility.runWorkUnderContext(PartitionUtility.java:41)
	at weblogic.work.SelfTuningWorkManagerImpl.runWorkUnderContext(SelfTuningWorkManagerImpl.java:617)
	at weblogic.work.ExecuteThread.execute(ExecuteThread.java:397)
	at weblogic.work.ExecuteThread.run(ExecuteThread.java:346)
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'encrypt-org.springframework.cloud.bootstrap.encrypt.KeyProperties': Initialization of bean failed; nested exception is java.lang.NoClassDefFoundError: org/springframework/security/rsa/crypto/RsaAlgorithm
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:591)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:502)
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:312)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:228)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:310)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:200)
	at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:251)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1138)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1065)
	at org.springframework.beans.factory.annotation.AutowiredAnnotationBeanPostProcessor$AutowiredFieldElement.inject(AutowiredAnnotationBeanPostProcessor.java:584)
	... 73 common frames omitted
Caused by: java.lang.NoClassDefFoundError: org/springframework/security/rsa/crypto/RsaAlgorithm
	at java.lang.Class.getDeclaredMethods0(Native Method)
	at java.lang.Class.privateGetDeclaredMethods(Class.java:2701)
	at java.lang.Class.getDeclaredMethods(Class.java:1975)
	at org.springframework.boot.context.properties.bind.JavaBeanBinder$Bean.putProperties(JavaBeanBinder.java:108)
	at org.springframework.boot.context.properties.bind.JavaBeanBinder$Bean.<init>(JavaBeanBinder.java:103)
	at org.springframework.boot.context.properties.bind.JavaBeanBinder$Bean.get(JavaBeanBinder.java:193)
	at org.springframework.boot.context.properties.bind.JavaBeanBinder.bind(JavaBeanBinder.java:48)
	at org.springframework.boot.context.properties.bind.Binder.lambda$null$5(Binder.java:339)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.ArrayList$ArrayListSpliterator.tryAdvance(ArrayList.java:1351)
	at java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:126)
	at java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:498)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:485)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:152)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:464)
	at org.springframework.boot.context.properties.bind.Binder.lambda$bindBean$6(Binder.java:340)
	at org.springframework.boot.context.properties.bind.Binder$Context.withIncreasedDepth(Binder.java:439)
	at org.springframework.boot.context.properties.bind.Binder$Context.withBean(Binder.java:425)
	at org.springframework.boot.context.properties.bind.Binder$Context.access$400(Binder.java:379)
	at org.springframework.boot.context.properties.bind.Binder.bindBean(Binder.java:337)
	at org.springframework.boot.context.properties.bind.Binder.bindObject(Binder.java:279)
	at org.springframework.boot.context.properties.bind.Binder.bind(Binder.java:221)
	at org.springframework.boot.context.properties.bind.Binder.lambda$bindBean$4(Binder.java:331)
	at org.springframework.boot.context.properties.bind.JavaBeanBinder.bind(JavaBeanBinder.java:72)
	at org.springframework.boot.context.properties.bind.JavaBeanBinder.bind(JavaBeanBinder.java:61)
	at org.springframework.boot.context.properties.bind.JavaBeanBinder.bind(JavaBeanBinder.java:53)
	at org.springframework.boot.context.properties.bind.Binder.lambda$null$5(Binder.java:339)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.ArrayList$ArrayListSpliterator.tryAdvance(ArrayList.java:1351)
	at java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:126)
	at java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:498)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:485)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:152)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:464)
	at org.springframework.boot.context.properties.bind.Binder.lambda$bindBean$6(Binder.java:340)
	at org.springframework.boot.context.properties.bind.Binder$Context.withIncreasedDepth(Binder.java:439)
	at org.springframework.boot.context.properties.bind.Binder$Context.withBean(Binder.java:425)
	at org.springframework.boot.context.properties.bind.Binder$Context.access$400(Binder.java:379)
	at org.springframework.boot.context.properties.bind.Binder.bindBean(Binder.java:337)
	at org.springframework.boot.context.properties.bind.Binder.bindObject(Binder.java:279)
	at org.springframework.boot.context.properties.bind.Binder.bind(Binder.java:221)
	at org.springframework.boot.context.properties.bind.Binder.bind(Binder.java:210)
	at org.springframework.boot.context.properties.bind.Binder.bind(Binder.java:192)
	at org.springframework.boot.context.properties.ConfigurationPropertiesBinder.bind(ConfigurationPropertiesBinder.java:78)
	at org.springframework.boot.context.properties.ConfigurationPropertiesBindingPostProcessor.bind(ConfigurationPropertiesBindingPostProcessor.java:106)
	at org.springframework.boot.context.properties.ConfigurationPropertiesBindingPostProcessor.postProcessBeforeInitialization(ConfigurationPropertiesBindingPostProcessor.java:93)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.applyBeanPostProcessorsBeforeInitialization(AbstractAutowireCapableBeanFactory.java:423)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1702)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:583)
	... 82 common frames omitted
Caused by: java.lang.ClassNotFoundException: org.springframework.security.rsa.crypto.RsaAlgorithm
	at weblogic.utils.classloaders.GenericClassLoader.findLocalClass(GenericClassLoader.java:981)
	at weblogic.utils.classloaders.GenericClassLoader.findClass(GenericClassLoader.java:942)
	at weblogic.utils.classloaders.ChangeAwareClassLoader.findClass(ChangeAwareClassLoader.java:83)
	at weblogic.utils.classloaders.GenericClassLoader.doFindClass(GenericClassLoader.java:564)
	at weblogic.utils.classloaders.GenericClassLoader.loadClass(GenericClassLoader.java:500)
	at weblogic.utils.classloaders.GenericClassLoader.loadClass(GenericClassLoader.java:464)
	at weblogic.utils.classloaders.GenericClassLoader.loadClass(GenericClassLoader.java:441)
	at weblogic.utils.classloaders.ChangeAwareClassLoader.loadClass(ChangeAwareClassLoader.java:53)
	... 135 common frames omitted
```

**Comment by wilkinsona** [comment]
`JpaProperties` should only be enabled when `HibernateJpaAutoConfiguration` is enabled. `HibernateJpaAutoConfiguration` should only be enabled when `org.hibernate.ejb.HibernateEntityManager` or `org.hibernate.jpa.HibernateEntityManager` is on the class path. I wonder if WebLogic is polluting your application's class path with an old version of Hibernate that has one or both of these classes but does not have `org.hibernate.boot.model.naming.ImplicitNamingStrategy`.

**Comment by aritzbastida** [comment]
True enough. We are using JpaProperties and JpaBaseConfiguration as a basis for our own EclipselinkJpaAutoConfiguration class; however, you are right, it has a strong dependency with Hibernate, so I might need to start from scratch (maybe copy paste the necessary parts from JpaBaseConfiguration; I tried to use most of the infrastructure as possible).

However, the question may be clearer in the second example: KeyProperties is bound in both cases because of a positive match of EncryptionBootstrapConfiguration. However, in standalone, the code in JavaBeanBinder.putProperties(JavaBeanBinder.java:108) won't be called; in Weblogic, it does. I was wondering if the binding process might be different...

**Comment by wilkinsona** [comment]
Thanks. We're aware of the oddities of `JpaProperties` and plan to revisit it in 2.1 (see #11112).

The second example looks like a bug in Spring Cloud to me. Its public `getRsa()` method returns an instance of `KeyProperties.Rsa`. The `getAlgorithm()` method on `KeyProperties.Rsa` returns a `org.springframework.security.rsa.crypto.RsaAlgorithm`. Depending on the JVM that's being used and its verifier, this link from `KeyProperties` to `RsaAlgorithm` will cause a class loading failure.

I can reproduce the failure that you are seeing (which is different to the failure I've described above) outside of WebLogic by excluding `org.springframework.security:spring-security-rsa` and adding `encrypt.rsa.strong=true` to `application.properties`. I've opened https://github.com/spring-cloud/spring-cloud-commons/issues/334 as I think this should be addressed in Spring Cloud.
